### PR TITLE
decasify: init at 0.8.0

### PR DESCRIPTION
**Pull request by alerque**
New package for [decasify](https://github.com/alerque/decasify). For now this is just packaging the CLI tool. In the future there is a possibility of packaging some of the other iterations of tooling in the same repository. For example the Lua module may become a dependency for sile. Also there is a NeoVIM plugin as well.

The package derivation is almost identical for the one for [git-warp-time](https://github.com/alerque/git-warp-time) recently merged in #354046.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by getchoo — CHANGES_REQUESTED** [review]

File: pkgs/development/lua-modules/generated-packages.nix
Comment (alerque) [reply pkgs/development/lua-modules/generated-packages.nix:1]: Yuck, that second commit wasn't ready or tested yet, I didn't mean to push it here at all, I meant to push it to another machine of mine. Sorry, I just backed it out and we're back to the original this PR was targeting.

File: pkgs/by-name/de/decasify/package.nix
Comment (getchoo) [inline pkgs/by-name/de/decasify/package.nix:24]:
```suggestion
    hash = "sha256-HTUAb/yL3H4B/n/Ecd/fDpnTYiqwco/E07sa6pFIIU4=";
```

File: pkgs/by-name/de/decasify/package.nix
Comment (getchoo) [inline pkgs/by-name/de/decasify/package.nix:28]:
```suggestion
    inherit (finalAttrs) pname version src;
```

File: pkgs/by-name/de/decasify/package.nix
Comment (getchoo) [inline pkgs/by-name/de/decasify/package.nix:30]: I can't say this is something I see usually...is there any specific reason for it? If so, a comment would be nice

File: pkgs/by-name/de/decasify/package.nix
Comment (getchoo) [inline pkgs/by-name/de/decasify/package.nix:35]: This would also usually go into buildInputs; is it something required for the build?

File: pkgs/by-name/de/decasify/package.nix
Comment (getchoo) [inline pkgs/by-name/de/decasify/package.nix:46]:
```suggestion
```
This seems to only be called by `git-cliff`, which shouldn't be happening when we're only building. If it is required though, it should be in `nativeBuildInputs` so that we can ensure the system building this can actually execute it

**Review by getchoo — CHANGES_REQUESTED** [review]

File: pkgs/by-name/de/decasify/package.nix
Comment (alerque) [reply pkgs/by-name/de/decasify/package.nix:46]: Sorry this was leftover from work for a dev shell in the upstream flake, removing.

File: pkgs/by-name/de/decasify/package.nix
Comment (alerque) [reply pkgs/by-name/de/decasify/package.nix:28]: This change causes the build to fail.

File: pkgs/by-name/de/decasify/package.nix
Comment (alerque) [reply pkgs/by-name/de/decasify/package.nix:27]: A comment might be appropriate, but perhaps @doronbehar could better explain why this is the way it is. This is a solution we worked out for the `git-warp-time` derivation and also applied to `sile`. I just know it doesn't work right without it. It seems to be a bug in the Nix build system actually because ZIP or GZIP compressed sources are handled differently. The extraction from `fetchurl` happens more or less automatically with the right decompression either way, but `rustPlatform.fetchCargoTarball` doesn't catch a clue and follow suite. If you happen to have downloaded and cached the upstream sources already (e.g. fetched locally to figure out the hash) then a local build will work fine without this bit, but it will fail remotely in CI where the matching hash artifact doesn't exist *before* the build.

File: pkgs/by-name/de/decasify/package.nix
Comment (alerque) [reply pkgs/by-name/de/decasify/package.nix:32]: The upstream canonical sources and ZSTD compressed, which is not quite handled automatically the way GZIP or other compressions are in Nix. See the [other comment](https://github.com/NixOS/nixpkgs/pull/355894#discussion_r1864138388) on this topic.

**Comment by alerque** [comment]
I went ahead and applied some of the review comments here to `git-warp-time` which had a similarly unnecessary `typos` in the buildInputs.

File: pkgs/by-name/de/decasify/package.nix
Comment (doronbehar) [reply pkgs/by-name/de/decasify/package.nix:25]: @alerque adding `pname` and `version` there has the small benefit of a small benefit of producing a slightly more informative `$out` path:

```
/nix/store/x2l73974ganpvd79sdwfl0b5z1gjvlkk-decasify-0.8.0-vendor.tar.gz
```

Instead of:

```
/nix/store/c5lj5pl5ladzb9yw6mg6gkabh8xzyh6r-cargo-deps-vendor.tar.gz
```

The unfortunate detail that made this change "fail" for you, is the fact that this causes a change of hash, see:

- https://github.com/NixOS/nixpkgs/issues/85703

This worked for me locally:

```nix
  cargoDeps = rustPlatform.fetchCargoTarball {
    inherit (finalAttrs) pname version src;
    nativeBuildInputs = [ zstd ];
    dontConfigure = true;
    hash = "sha256-bD8MYufI87j//7dIAnCzmp4yoOaT81Zv1i7rjWpjPlc=";
  };
```

File: pkgs/by-name/de/decasify/package.nix
Comment (doronbehar) [reply pkgs/by-name/de/decasify/package.nix:27]: > but perhaps @doronbehar could better explain why this is the way it is

Yes this is slightly tricky. It seems that by default the derivation that the derivation `fetchCargoTarball` builds (defined [here](https://github.com/NixOS/nixpkgs/blob/e405f30513169feedb64b5c25e7b00242010af58/pkgs/build-support/rust/fetch-cargo-tarball/default.nix#L72-L157)), tries to run the default `./configure` phase that the Nix `stdenv` tries to run. Usually for most Rust projects, there is no `./configure` script and no body has to specify that, but here and both in `sile` and `git-warp-time`, the `src` contains a `./configure` script that shouldn't be run if we just want to prefetch the rust crates...

I wouldn't consider this a bug, but perhaps the default behavior should be changed. And I agree that it would be nice to explain this in a comment, even a short one:

```
# so the cargo fetcher won't try to run the `./configure` script
```

Comment (alerque) [reply pkgs/by-name/de/decasify/package.nix:27]: Aah, that's making more sense. In that case yes I would definitely suggest that a command `fetchCargTarball` shouldn't be running a configure phase even if it exists. I understand my use of GNU Autotools with Rust projects is a little unusual, but unusual is not wrong and on the flip side I can't think of a single case where a configure script of any kind should be taken into account when figuring the contents of a Cargo sources vendor tarball. In fact in pure Cargo based projects this is never the case either, `cargo fetch` specifically does *not* take into account the current architecture or target triple or any features (all, default, no-std, etc.), it is expected to fetch all possible permutations of dependencies and sort our what gets *used* later in the `cargo build` or `cargo test` or whatever stages when feature flags, platform targets, etc. *can be * specified.

File: pkgs/by-name/de/decasify/package.nix
Comment (alerque) [reply pkgs/by-name/de/decasify/package.nix:25]: Whose idea was it to make the final filename part of the hash (but only part of the file name because of course the hash itself is also in the file name)? On second thought maybe don't tell me, I might be inclined to take `git blame` literally ;-P

File: pkgs/by-name/de/decasify/package.nix
Comment (doronbehar) [reply pkgs/by-name/de/decasify/package.nix:27]: I agree: https://github.com/NixOS/nixpkgs/pull/360540

**Comment by alerque** [comment]
I've now appended another commit because the `sile` derivation had two of the same issues brought up in this review: the cargo vendor tarball naming and needless dependencies that are developer-only tooling and not used to build or run the app.

**Comment by doronbehar** [comment]
There is a merge conflict with `sile`...

**Comment by doronbehar** [comment]
@getchoo please see the changes applied and unblock the PR :).

**Review by getchoo — APPROVED** [review]
Built and seems to work fine on x86_64-linux. The `sile` build failure reported by Ofborg is also [already present on master](https://hydra.nixos.org/build/279510156)

File: pkgs/by-name/de/decasify/package.nix
Comment (doronbehar) [reply pkgs/by-name/de/decasify/package.nix:49]: I don't understand why people use `nix-update-script` for simple packages that use standard GitHub tags / releases - how does that makes a difference for the common updating scripts like `nix-update` and `nixpkgs-update`?

Comment (doronbehar) [reply pkgs/by-name/de/decasify/package.nix:49]: > > I don't understand why people use nix-update-script for simple packages that use standard GitHub tags / releases - how does that makes a difference for the common updating scripts like nix-update and nixpkgs-update?
> 
> The biggest reason would be that it allows the package to be updated with [./maintainers/scripts/update.nix](https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/update.nix?rgh-link-date=2024-12-02T20%3A34%3A24Z). It also usually results in more consistent updates from nixpkgs-review/r-ryantm as you're not longer depending on other repositories to update the package with Repology, or nixpkgs-review supporting other attributes you use like `cargoHash`, `pnpmDeps`, etc. (`nix-update` is much better at this)

Evidently `nixpkgs-update` is updating packages based on GitHub releases, and for sure it has no issues with Rust packages that has a `cargoHash`. Perhaps it doesn't support updating hashes of `pnpmDeps`, but that is not an issue for this package. That `update.nix` script is pretty good, but still since `nix-update` is better I'd still not bother adding an `updateScript` for a trivial case like this.

> Edit: Also I'm not really sure why this PR was merged so soon after this approval...probably should've given OP a chance to add this if they wanted to, _especially_ the test

The `passthru.tests.version` attribute is not so bad, but for classic standard packages like this [`versioncheckhook`](https://nixos.org/manual/nixpkgs/stable/#versioncheckhook) is better, as also [claimed by the docs](https://nixos.org/manual/nixpkgs/unstable/#var-passthru-tests-packages), see also a slightly thorough comparison being worked on [here](https://github.com/NixOS/nixpkgs/pull/348025). Due to that, I thought it is not worth it to wait for 1+ more day for CI just for adding these attributes. The maintainer of course is welcome to do any of the above in a separate PR.

**Pull request merged**
